From 0e6bee4574fe6833713bbe52dd5cd6cfaeecd8f7 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:38 -0400
Subject: [net] netfilter: nf_tables: Avoid duplicate call to nft_data_uninit() for same key

Message-id: <20141021161937.22888.73652.stgit@localhost.localdomain>
Patchwork-id: 97982
O-Subject: [RHEL7 net PATCH 085/124] netfilter: nf_tables: Avoid duplicate call to nft_data_uninit() for same key
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 0dc1362562a2e8b82a6be8d3ae307a234f28f9bc

 commit 0dc1362562a2e8b82a6be8d3ae307a234f28f9bc
 Author: Thomas Graf <tgraf@suug.ch>
 Date:   Fri Aug 1 17:25:38 2014 +0200

    netfilter: nf_tables: Avoid duplicate call to nft_data_uninit() for same key

    nft_del_setelem() currently calls nft_data_uninit() twice on the same
    key. Once to release the key which is guaranteed to be NFT_DATA_VALUE
    and a second time in the error path to which it falls through.

    The second call has been harmless so far though because the type
    passed is always NFT_DATA_VALUE which is currently a no-op.

    Signed-off-by: Thomas Graf <tgraf@suug.ch>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 119de56..1b0fdf9 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3144,6 +3144,7 @@ static int nft_del_setelem(struct nft_ctx *ctx, struct nft_set *set,
  if (set->flags & NFT_SET_MAP)
   nft_data_uninit(&elem.data, set->dtype);
 
+ return 0;
 err2:
  nft_data_uninit(&elem.key, desc.type);
 err1:
-- 
1.7.1